File: c#/iterator/Ejm1/ConcreteIterator.cs

namespace Iterator.Ejm1
{
    /// <summary>
    /// Implementa la interface Iterator.
    /// Mantiene un registro de la posición actual en el recorrido del agregado (todo)
    /// </summary>
    class ConcreteIterator : Iterator
    {
        private ConcreteAggregate aggregate;
        private int current = 0;

        // Constructor 
        public ConcreteIterator(ConcreteAggregate aggregate)
        {
            this.aggregate = aggregate;
        }

        public override object First()
        {
            return aggregate[0];
        }

        public override object Next()
        {
            object ret = null;
            if (current < aggregate.Count - 1)
            {
                ret = aggregate[++current];
            }

            return ret;
        }

        public override object CurrentItem()
        {
            return aggregate[current];
        }

        public override bool IsDone()
        {
            return current >= aggregate.Count ? true : false;
        }
    }
}

File: java/iterator/ej1/ConcreteIterator.java
package patrones.iterator.ej1;

public class ConcreteIterator extends Iterator
{
    private ConcreteAggregate aggregate;
    private int current = 0;

    // Constructor 
    public ConcreteIterator(ConcreteAggregate aggregate)
    {
        this.aggregate = aggregate;
    }

    public Object first()
    {
        return aggregate.getItem(0);
    }

    public Object next()
    {
        Object ret = null;
        if (current < aggregate.count() - 1)
        {
            ret = aggregate.getItem(++current);
        }

        return ret;
    }

    public Object currentItem()
    {
        return aggregate.getItem(current);
    }

    public boolean isDone()
    {
        if (current >= aggregate.count())
        	return true;
        else
        	return false;
        			
    }
}
